src/monster.cpp

```diff
@@ -1173,6 +1173,28 @@ void monster::melee_attack( Creature &target, bool, const matec_id&, int hitspre
     }
 }
 
+void monster::on_damage_taken( Creature *source, dealt_damage_instance &d ) // trigger effects based on dmg type/amount
+{
+    if ( has_flag(MF_ACID_BLOOD )) { // currently only spreading acid blood, can be used for any on-hit splatter
+        if ( rng(0, 1) ) { // you dont hit a cyst, blood vessel every time     
+            const tripoint origin = pos();
+            const int totalDealt = d.type_damage(DT_CUT) + d.type_damage(DT_STAB) + floor(d.type_damage(DT_BASH) * 0.5);
+            // Bashing damage -> less chance to spread acid
+            // debugmsg("on_dmg scope target totalDealt %d", totalDealt);
+
+            const float fraction = float(totalDealt) / get_hp_max(); // % of total hp this hit takes off
+            const int radius = rng(1, 2);
+            const int density = 1;
+            const int amount = ceil(fraction * 5); // 1 acid for each 20 % damage this takes off maxHP, min 1
+
+            for ( int i = 0; i < amount; i++ ) {
+                const tripoint dest( origin.x + rng(-radius, radius), origin.y + rng(-radius, radius), origin.z );
+                g->m.add_field( dest, bloodType(), density );
+            }
+        }
+    }
+}
```

Comment:
The whole function just replicates the functionality of `mdefense::acidsplash` except with fields.
Having a totally separate function that hardcoded some behavior incredibly similar to existing one is really bad practice.

There exists a problem with monster defense functions being called before an attack rather than after it, but this is easily fixed by moving them downward in just a handful of functions.

You should:
* Supply `const dealt_damage_instance &dam` or `const dealt_damage_instance * const dam` to monster defenses. Search for `mon_action_defend` and `sp_defense` - literally only a handful
* Move `on_hit` below `deal_damage` in `Creature::deal_melee_hit`, so that it can get the new argument
* Rewrite this little block here as a proper monster defense
* (Optional) Add some temporary overload (with clear warning in comment) to handle `on_hit` without either the projectile argument or the new argument. Alternatively, update all `on_hit` (there is quite a bunch of them)

Reply:
I can see what you are saying here. However, the functions in question are true defense mechanicsm, as in zapping, or or spraying acid on a potential attacker.
What im trying to do here is add some common sense to acid-type zombies. If they receive a wound, they bleed acid, just like a human would lose blood after being slashed.

Basicly what im thinking is that my function covers a different gameplay mechanic than mdefense. One is explicitly used to damage an attacker, while the other is MEANT for immersion purpose mostly.

The mdefense::acidsplash functionality is a unique trait of the corrosive zombie, while on_damage_taken is a generic function and not necesarrily limited on Acidblood-flag.
For example, units without Acidblood-flag could be used to spill normal blood after being damaged, hence "on_damage_taken".

Does anything im saying here sway your opinion ?

Reply:
Adding "blood splash" defense to monsters is easy and could be extended to other types (particularly juicy triffid, for example).
A hardcoded behavior that explicitly targets only acid blood type monsters, can't be overriden, can't be extended, uses its own code sections not shared with other "on_monster_got_hit_by_something" types, breaks the definition of `on_hurt` (`on_some_hurts` is not exactly intuitive), is a burden on maintainers and contributors.

And there is a problem of immersion itself: acid field is created when a puddle of acid is big and potent enough that you can't really stand over it without standing in it and getting burned. Acid spills comparable in size to blood spills from knife wounds wouldn't create anything of this magnitude, meaning that you'll need to use gameplay reasons to justify this.